src/main/java/com/jcraft/jsch/Session.java

```diff
@@ -81,7 +81,8 @@
   private static final int PACKET_MAX_SIZE = 256 * 1024;
 
   private byte[] V_S;                                 // server version
-  private byte[] V_C=Util.str2byte("SSH-2.0-JSCH_"+JSch.VERSION); // client version
+  private byte[] V_C; // will be set during connect so that the client version
```

Comment:
Why not simply do `VC=Util.str2byte(jsch.getClientVersion())` here?
Then we could omit all the other changes to this class.

Reply:
There might be a setClientVersion on Jsch-level after creating a session and before doing a connect that would get lost that way

Reply:
I think it's easier to document and understand that calls to `jsch.setClientVersion()` have no impact to any previously created Session instances than it is to document and understand that `jsch.setClientVersion()` affects any previously created Session instance that has not yet execute `session.connect()`.

Basically: after you create a Session object, via `jsch.getSession()`, the only way to manipulate the client version string of said Session object should be via `session.setClientVersion()`, independent of any calls to `jsch.setClientVersion()` that occur after you have created the Session object.

Reply:
Even if we implement it the way you suggested you need to either have a distinct member for the client version bytes used during a connection or check the state of the session and throw an exception if you try to set the version string after a connection has been established.

Otherwise you will run into the problem you've described before:

> If a user changed the client version string after a Session is connected and a rekeying
> operation occurred, their SSH connection would then break, because the client version
> is supposed to be idempotent.

Personally I don't think that this is hard to document and I can add a javadoc comment trying to do it to the method (I would have done so anyway, if the method didn't exist already).

Reply:
Added some javadoc that should explain the way this works.

Reply:
I still disagree with this.
@mwiede what are your thoughts?

Reply:
@mwiede Any thoughts on this?
Should we just go with a simpler approach I suggested or the more complex approach implemented here?